File: packages/py/cairn-p2p/src/cairn/transport/chain.py
"""Transport abstraction and 9-level fallback chain."""

from __future__ import annotations

import asyncio
from abc import ABC, abstractmethod
from dataclasses import dataclass
from enum import Enum, auto
from typing import Any, Callable, Coroutine

DEFAULT_TIMEOUT: float = 10.0


class TransportType(Enum):
    """Transport type in the 9-level fallback chain."""

    DIRECT_QUIC = auto()
    STUN_HOLE_PUNCH = auto()
    DIRECT_TCP = auto()
    TURN_UDP = auto()
    TURN_TCP = auto()
    WEBSOCKET_TLS = auto()
    WEBTRANSPORT_H3 = auto()
    CIRCUIT_RELAY_V2 = auto()
    HTTPS_LONG_POLL = auto()

    @property
    def priority(self) -> int:
        """Priority number (1 = best, 9 = worst)."""
        return list(TransportType).index(self) + 1

    @property
    def tier0_available(self) -> bool:
        """Whether this transport is available in zero-config mode."""
        return self in (
            TransportType.DIRECT_QUIC,
            TransportType.STUN_HOLE_PUNCH,
            TransportType.DIRECT_TCP,
            TransportType.CIRCUIT_RELAY_V2,
        )

    @property
    def display_name(self) -> str:
        """Human-readable name."""
        names = {
            TransportType.DIRECT_QUIC: "Direct QUIC v1",
            TransportType.STUN_HOLE_PUNCH: "STUN-assisted UDP",
            TransportType.DIRECT_TCP: "Direct TCP",
            TransportType.TURN_UDP: "TURN relay (UDP)",
            TransportType.TURN_TCP: "TURN relay (TCP)",
            TransportType.WEBSOCKET_TLS: "WebSocket/TLS (443)",
            TransportType.WEBTRANSPORT_H3: "WebTransport/HTTP3",
            TransportType.CIRCUIT_RELAY_V2: "Circuit Relay v2",
            TransportType.HTTPS_LONG_POLL: "HTTPS long-polling",
        }
        return names[self]

    @classmethod
    def all_in_order(cls) -> list[TransportType]:
        """All transport types in priority order."""
        return list(cls)


class Transport(ABC):
    """Abstract base class for transport implementations."""

    @abstractmethod
    async def connect(
        self, addr: str, timeout: float = DEFAULT_TIMEOUT
    ) -> None:
        """Connect to a remote address."""

    @abstractmethod
    async def send(self, data: bytes) -> None:
        """Send data to the connected peer."""

    @abstractmethod
    async def receive(self) -> bytes:
        """Receive data from the connected peer."""

    @abstractmethod
    async def close(self) -> None:
        """Close the transport connection."""


@dataclass
class TransportAttempt:
    """Configuration for a single transport attempt."""

    transport_type: TransportType
    timeout: float = DEFAULT_TIMEOUT
    available: bool = True

    @property
    def priority(self) -> int:
        return self.transport_type.priority


@dataclass
class TransportAttemptResult:
    """Result of attempting a single transport."""

    transport_type: TransportType
    error: str | None = None
    skipped: bool = False
    duration: float = 0.0

    def __str__(self) -> str:
        if self.skipped:
            return (
                f"{self.transport_type.display_name}: "
                f"skipped (not configured)"
            )
        if self.error:
            return (
                f"{self.transport_type.display_name}: "
                f"failed ({self.error}) [{self.duration:.3f}s]"
            )
        return (
            f"{self.transport_type.display_name}: "
            f"success [{self.duration:.3f}s]"
        )


class TransportExhaustedError(Exception):
    """All transports in the fallback chain failed."""

    def __init__(
        self,
        details: str,
        suggestion: str,
        results: list[TransportAttemptResult],
    ) -> None:
        self.details = details
        self.suggestion = suggestion
        self.results = results
        super().__init__(
            f"all transports exhausted: {details}. {suggestion}"
        )


class FallbackChain:
    """Executes the 9-level transport priority chain.

    Supports both sequential and parallel (ICE-style) probing modes.
    """

    def __init__(
        self,
        per_transport_timeout: float = DEFAULT_TIMEOUT,
        has_turn: bool = False,
        has_relay_443: bool = False,
        parallel_mode: bool = False,
    ) -> None:
        self._parallel_mode = parallel_mode
        self._transports: list[TransportAttempt] = []

        for tt in TransportType.all_in_order():
            if tt in (TransportType.TURN_UDP, TransportType.TURN_TCP):
                available = has_turn
            elif tt in (
                TransportType.WEBSOCKET_TLS,
                TransportType.WEBTRANSPORT_H3,
                TransportType.HTTPS_LONG_POLL,
            ):
                available = has_relay_443
            else:
                available = True

            self._transports.append(
                TransportAttempt(
                    transport_type=tt,
                    timeout=per_transport_timeout,
                    available=available,
                )
            )

    @classmethod
    def tier0(
        cls, per_transport_timeout: float = DEFAULT_TIMEOUT
    ) -> FallbackChain:
        """Create a Tier 0 (zero-config) fallback chain."""
        return cls(per_transport_timeout, False, False, False)

    @property
    def transports(self) -> list[TransportAttempt]:
        return list(self._transports)

    @property
    def parallel_mode(self) -> bool:
        return self._parallel_mode

    async def execute(
        self,
        attempt_fn: Callable[
            [TransportType, float],
            Coroutine[Any, Any, Any],
        ],
    ) -> tuple[TransportType, Any]:
        """Execute the fallback chain.

        attempt_fn(transport_type, timeout) should return the result
        on success or raise an exception on failure.

        Returns (transport_type, result) for the first success.
        Raises TransportExhaustedError if all fail.
        """
        if self._parallel_mode:
            return await self._execute_parallel(attempt_fn)
        return await self._execute_sequential(attempt_fn)

    async def _execute_sequential(
        self,
        attempt_fn: Callable[
            [TransportType, float],
            Coroutine[Any, Any, Any],
        ],
    ) -> tuple[TransportType, Any]:
        results: list[TransportAttemptResult] = []

        for attempt in self._transports:
            if not attempt.available:
                results.append(
                    TransportAttemptResult(
                        transport_type=attempt.transport_type,
                        skipped=True,
                    )
                )
                continue

            loop = asyncio.get_event_loop()
            start = loop.time()
            try:
                value = await attempt_fn(
                    attempt.transport_type, attempt.timeout
                )
                return (attempt.transport_type, value)
            except Exception as e:
                elapsed = loop.time() - start
                results.append(
                    TransportAttemptResult(
                        transport_type=attempt.transport_type,
                        error=str(e),
                        duration=elapsed,
                    )
                )

        raise _build_exhausted_error(results)

    async def _execute_parallel(
        self,
        attempt_fn: Callable[
            [TransportType, float],
            Coroutine[Any, Any, Any],
        ],
    ) -> tuple[TransportType, Any]:
        results: list[TransportAttemptResult] = []
        tasks: dict[asyncio.Task[Any], TransportType] = {}

        for attempt in self._transports:
            if not attempt.available:
                results.append(
                    TransportAttemptResult(
                        transport_type=attempt.transport_type,
                        skipped=True,
                    )
                )
                continue

            task = asyncio.create_task(
                attempt_fn(attempt.transport_type, attempt.timeout)
            )
            tasks[task] = attempt.transport_type

        if not tasks:
            raise _build_exhausted_error(results)

        pending = set(tasks.keys())

        while pending:
            done, pending = await asyncio.wait(
                pending, return_when=asyncio.FIRST_COMPLETED
            )

            for task in done:
                tt = tasks[task]
                exc = task.exception()
                if exc is None:
                    # Cancel remaining tasks
                    for p in pending:
                        p.cancel()
                    return (tt, task.result())
                else:
                    results.append(
                        TransportAttemptResult(
                            transport_type=tt,
                            error=str(exc),
                        )
                    )

        raise _build_exhausted_error(results)


def _build_exhausted_error(
    results: list[TransportAttemptResult],
) -> TransportExhaustedError:
    details = "; ".join(str(r) for r in results)
    has_skipped = any(r.skipped for r in results)
    suggestion = (
        "deploy companion infrastructure (TURN relay, "
        "WebSocket relay on port 443) to enable additional "
        "transport fallbacks"
        if has_skipped
        else "check network connectivity and firewall rules"
    )
    return TransportExhaustedError(details, suggestion, results)

File: packages/go/cairn-p2p/transport/chain.go
package transport

import (
	"context"
	"fmt"
	"sync"
	"time"

	cairn "github.com/moukrea/cairn/packages/go/cairn-p2p"
	"github.com/moukrea/cairn/packages/go/cairn-p2p/crypto"
)

// TransportType represents a transport in the 9-level fallback chain.
type TransportType int

const (
	TransportDirectQUIC     TransportType = iota + 1 // Priority 1: Direct UDP (QUIC v1, RFC 9000)
	TransportSTUNHolePunch                            // Priority 2: STUN-assisted UDP hole punch
	TransportDirectTCP                                // Priority 3: Direct TCP
	TransportTURNUDP                                  // Priority 4: TURN relay (UDP)
	TransportTURNTCP                                  // Priority 5: TURN relay (TCP)
	TransportWebSocket                                // Priority 6: WebSocket over TLS (port 443)
	TransportWebTransport                             // Priority 7: WebTransport over HTTP/3 (port 443)
	TransportCircuitRelayV2                           // Priority 8: Circuit Relay v2 (transient)
	TransportHTTPSPolling                             // Priority 9: HTTPS long-polling (port 443)
)

// String returns a human-readable name for the transport type.
func (t TransportType) String() string {
	switch t {
	case TransportDirectQUIC:
		return "direct-quic"
	case TransportSTUNHolePunch:
		return "stun-hole-punch"
	case TransportDirectTCP:
		return "direct-tcp"
	case TransportTURNUDP:
		return "turn-udp"
	case TransportTURNTCP:
		return "turn-tcp"
	case TransportWebSocket:
		return "websocket"
	case TransportWebTransport:
		return "webtransport"
	case TransportCircuitRelayV2:
		return "circuit-relay-v2"
	case TransportHTTPSPolling:
		return "https-polling"
	default:
		return fmt.Sprintf("unknown(%d)", int(t))
	}
}

// Tier0Transports are transports available at Tier 0 (no companion infrastructure).
var Tier0Transports = []TransportType{
	TransportDirectQUIC,
	TransportSTUNHolePunch,
	TransportDirectTCP,
}

// AllTransports is the full 9-level transport priority list.
var AllTransports = []TransportType{
	TransportDirectQUIC,
	TransportSTUNHolePunch,
	TransportDirectTCP,
	TransportTURNUDP,
	TransportTURNTCP,
	TransportWebSocket,
	TransportWebTransport,
	TransportCircuitRelayV2,
	TransportHTTPSPolling,
}

const (
	// DefaultTransportTimeout is the default per-transport connection timeout.
	DefaultTransportTimeout = 10 * time.Second
)

// DefaultStunServers are public STUN servers used for NAT type detection at Tier 0.
var DefaultStunServers = []string{
	"stun.l.google.com:19302",
	"stun1.l.google.com:19302",
	"stun.cloudflare.com:3478",
}

// TransportConfig holds configuration for the transport chain.
type TransportConfig struct {
	// TransportTimeout is the per-transport connection timeout (default 10s).
	TransportTimeout time.Duration

	// StunServers for NAT type detection.
	StunServers []string

	// TurnServers for relay transports (priorities 4-5).
	TurnServers []TurnServer

	// EnabledTransports controls which transports to attempt (default: all).
	EnabledTransports []TransportType
}

// TurnServer holds TURN relay server configuration.
type TurnServer struct {
	URL      string
	Username string
	Password string
}

// DefaultTransportConfig returns a TransportConfig with default values.
func DefaultTransportConfig() *TransportConfig {
	return &TransportConfig{
		TransportTimeout:  DefaultTransportTimeout,
		StunServers:       DefaultStunServers,
		EnabledTransports: AllTransports,
	}
}

// TransportAttemptResult records the outcome of attempting a single transport.
type TransportAttemptResult struct {
	Transport TransportType
	Err       error
	Duration  time.Duration
}

// TransportProvider is the interface that individual transport implementations must satisfy.
type TransportProvider interface {
	// Type returns the transport type.
	Type() TransportType

	// Dial attempts to connect to a peer via this transport.
	Dial(ctx context.Context, peerID cairn.PeerID, addrs []string) error

	// IsAvailable reports whether this transport can be attempted in the current environment.
	IsAvailable() bool
}

// TransportChain manages the 9-level transport fallback chain with parallel ICE-style probing.
type TransportChain struct {
	mu sync.Mutex

	identity  *crypto.IdentityKeypair
	config    *TransportConfig
	providers map[TransportType]TransportProvider
}

// NewTransportChain creates a new transport chain with the given identity and configuration.
func NewTransportChain(identity *crypto.IdentityKeypair, config *TransportConfig) *TransportChain {
	if config == nil {
		config = DefaultTransportConfig()
	}
	return &TransportChain{
		identity:  identity,
		config:    config,
		providers: make(map[TransportType]TransportProvider),
	}
}

// RegisterProvider registers a transport provider for a specific transport type.
func (tc *TransportChain) RegisterProvider(provider TransportProvider) {
	tc.mu.Lock()
	defer tc.mu.Unlock()
	tc.providers[provider.Type()] = provider
}

// Connect attempts to connect to a peer using the fallback chain with parallel probing.
// Returns the transport type that succeeded, or a TransportExhausted error with details.
func (tc *TransportChain) Connect(ctx context.Context, peerID cairn.PeerID, addrs []string) (TransportType, error) {
	tc.mu.Lock()
	enabled := tc.config.EnabledTransports
	timeout := tc.config.TransportTimeout
	providers := make(map[TransportType]TransportProvider)
	for k, v := range tc.providers {
		providers[k] = v
	}
	tc.mu.Unlock()

	type probeResult struct {
		transport TransportType
		err       error
		duration  time.Duration
	}

	resultCh := make(chan probeResult, len(enabled))

	probeCtx, cancel := context.WithCancel(ctx)
	defer cancel()

	// Launch parallel probes for all enabled transports that have providers
	launched := 0
	for _, tt := range enabled {
		provider, ok := providers[tt]
		if !ok || !provider.IsAvailable() {
			resultCh <- probeResult{
				transport: tt,
				err:       fmt.Errorf("transport %s: not available", tt),
			}
			launched++
			continue
		}

		launched++
		go func(p TransportProvider, t TransportType) {
			tCtx, tCancel := context.WithTimeout(probeCtx, timeout)
			defer tCancel()

			start := time.Now()
			err := p.Dial(tCtx, peerID, addrs)
			resultCh <- probeResult{
				transport: t,
				err:       err,
				duration:  time.Since(start),
			}
		}(provider, tt)
	}

	if launched == 0 {
		return 0, cairn.NewCairnError(
			cairn.ErrKindTransportExhausted,
			"all transports exhausted: no transports enabled",
			"Enable at least one transport or deploy companion infrastructure.",
		)
	}

	// Collect results; first success wins (ICE-style parallel probing)
	var failures []TransportAttemptResult
	for i := 0; i < launched; i++ {
		select {
		case r := <-resultCh:
			if r.err == nil {
				cancel() // cancel remaining probes
				return r.transport, nil
			}
			failures = append(failures, TransportAttemptResult{
				Transport: r.transport,
				Err:       r.err,
				Duration:  r.duration,
			})
		case <-ctx.Done():
			return 0, cairn.NewCairnError(
				cairn.ErrKindTransportExhausted,
				fmt.Sprintf("all transports exhausted: context cancelled after %d/%d attempts", len(failures), launched),
				"Check network connectivity and retry.",
			)
		}
	}

	// All transports failed — build detailed error
	return 0, newTransportExhaustedError(failures)
}

// Config returns the current transport configuration.
func (tc *TransportChain) Config() *TransportConfig {
	tc.mu.Lock()
	defer tc.mu.Unlock()
	return tc.config
}

// newTransportExhaustedError builds a detailed TransportExhausted error with per-transport details.
func newTransportExhaustedError(failures []TransportAttemptResult) *cairn.CairnError {
	details := ""
	for _, f := range failures {
		details += fmt.Sprintf("  %s: %v (%v)\n", f.Transport, f.Err, f.Duration)
	}

	suggestion := "Check network connectivity and retry."
	hasTier0Failure := false
	for _, f := range failures {
		switch f.Transport {
		case TransportDirectQUIC, TransportSTUNHolePunch, TransportDirectTCP:
			hasTier0Failure = true
		}
	}
	if hasTier0Failure && len(failures) <= 3 {
		suggestion = "Deploy companion TURN relay infrastructure to enable additional transports (priorities 4-9)."
	}

	return cairn.NewCairnError(
		cairn.ErrKindTransportExhausted,
		fmt.Sprintf("all transports exhausted (%d attempted):\n%s", len(failures), details),
		suggestion,
	)
}
